### ISSUE TITLE
ResourceOptions(parent=self) hangs pulumi preview indefinitely in CustomResource

### ISSUE THREAD

**Issue by rshade** [pulumi/pulumi#12032]
### What happened?

I have a CustomResource in python where when I use `parent=self`, it hangs the pulumi run indefinitely. The resource is defined as:
```
        self.namespace = Namespace(f"{self.name}-namespace",
            api_version="v1",
            kind="Namespace",
            metadata={"name": self.namespace_name},
            # opts=ResourceOptions(parent=self, depends_on=[cluster]),
            #opts=ResourceOptions(parent=self)
            opts=ResourceOptions(depends_on=[cluster])
        )
```

### Expected Behavior

It to proceed or an error being thrown for what is happening. 

### Steps to reproduce

Repro is available in slack. 

### Output of `pulumi about`

aws            5.27.0
cloudinit      1.3.0
datadog        4.12.0
google-native  0.27.0
kubernetes     3.23.1
python         unknown
random         4.8.2
tls            4.6.1

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by justinvp** [comment]
I was able to reproduce the issue with the provided repro from Slack, but I haven't had a chance to dive-in and debug yet.

**Comment by justinvp** [comment]
I've reduced the repro down to this:

```py
from typing import Optional

from pulumi import ComponentResource, ResourceOptions
from pulumi_random import RandomString


class First(ComponentResource):
    def __init__(self, name: str, opts: Optional[ResourceOptions]=None):
        super().__init__("pkg:index:first", name, {}, opts)


class Second(ComponentResource):
    def __init__(self, name: str, opts: Optional[ResourceOptions]=None):
        super().__init__("pkg:index:second", name, {}, opts)

        RandomString("str",
            length=10,
            opts=ResourceOptions(parent=self))


first = First("first")
second = Second("second",
    opts=ResourceOptions(parent=first, depends_on=[first]))
```

Either removing `parent=self` when creating the `RandomString`, or removing `depends_on=[first]` when creating the `Second` allows it to succeed. @rshade, I'm wondering if you can workaround by removing the `depends_on` in the similar place in your actual project. I'll sync with you on Slack.

**Comment by justinvp** [comment]
I was also able to use similar minimal repros in Node.js and Go. And I was able to address the problem in the same way (by removing the dependsOn).

## Node.js

```ts
import * as pulumi from "@pulumi/pulumi";
import * as random from "@pulumi/random";

class First extends pulumi.ComponentResource {
    constructor(name: string, opts?: pulumi.ResourceOptions) {
        super("pkg:index:first", name, {}, opts);
    }
}

class Second extends pulumi.ComponentResource {
    constructor(name: string, opts?: pulumi.ResourceOptions) {
        super("pkg:index:second", name, {}, opts);
        new random.RandomString("str", { length: 10 }, { parent: this });
    }
}

const first = new First("first");
const second = new Second("second", {
    parent: first,
    dependsOn: [first],
});
```

This didn't hang, but error'd with:

```
Diagnostics:
  pulumi:pulumi:Stack (fooparentts-dev):
    The Pulumi runtime detected that 61 promises were still active
    at the time that the process exited. There are a few ways that this can occur:
      * Not using `await` or `.then` on a Promise returned from a Pulumi API
      * Introducing a cyclic dependency between two Pulumi Resources
      * A bug in the Pulumi Runtime
    Leaving promises active is probably not what you want. If you are unsure about
    why you are seeing this message, re-run your program with the `PULUMI_DEBUG_PROMISE_LEAKS`
    environment variable. The Pulumi runtime will then print out additional
    debug information about the leaked promises.
```


## Go

```go
package main

import (
	"github.com/pulumi/pulumi-random/sdk/v4/go/random"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

type First struct {
	pulumi.ResourceState
}

func NewFirst(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) (*First, error) {
	res := &First{}
	err := ctx.RegisterComponentResource("pkg:index:first", name, res, opts...)
	if err != nil {
		return nil, err
	}
	return res, nil
}

type Second struct {
	pulumi.ResourceState
}

func NewSecond(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) (*Second, error) {
	res := &Second{}
	err := ctx.RegisterComponentResource("pkg:index:second", name, res, opts...)
	if err != nil {
		return nil, err
	}

	_, err = random.NewRandomString(ctx, "str", &random.RandomStringArgs{Length: pulumi.Int(10)}, pulumi.Parent(res))
	if err != nil {
		return nil, err
	}

	return res, nil
}

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		first, err := NewFirst(ctx, "first")
		if err != nil {
			return err
		}

		_, err = NewSecond(ctx, "second", pulumi.Parent(first), pulumi.DependsOn([]pulumi.Resource{first}))
		if err != nil {
			return err
		}

		return nil
	})
}
```

**Comment by justinvp** [comment]
It doesn’t appear to be an issue in the engine itself, as an equivalent program as an engine lifecycle test doesn’t hang: https://github.com/pulumi/pulumi/compare/master...justin/hangtest

My next step would be to debug into the Python SDK itself to understand why we’re hanging.

It seems like the repro is creating a cyclic dependency, so perhaps this is just unsupported, but I’m not sure how it worked in the past. If it is unsupported, we should emit an error rather than hanging.

**Comment by antifuchs** [comment]
So I just went through our code base and removed all instances of `depends_on=[that_parent]` which were unnecessary, and that does resolve our specific hang. 

I agree that it would be good to have a cycle detector for these cases, but am slightly worried about implicit dependencies: What happens when a resource uses another's outputs but is also a child of that second resource?

**Comment by justinvp** [comment]
OK, I've finally had a chance to dive in and I understand what's happening in Python:

Abhinav was kind enough to provide the following test (I've tweaked the variable names):

```python
@pulumi.runtime.test
def test_parent_and_depends_on_are_the_same_12032():
    mocks.set_mocks(MinimalMocks())

    first = pulumi.ComponentResource("pkg:index:first", "first")
    second = pulumi.ComponentResource(
        "pkg:index:second",
        "second",
        opts=pulumi.ResourceOptions(parent=first, depends_on=[first]),
    )

    # This would freeze before the fix.
    pulumi.CustomResource(
        "foo:bar:baz",
        "myresource",
        opts=pulumi.ResourceOptions(parent=second),
    )
```

Before calling `RegisterResource` for `second`, we expand its dependencies into a set of URNs via the following rules:

https://github.com/pulumi/pulumi/blob/1b2470db221167ec4e998e226b1c5338e1f26c7b/sdk/python/lib/pulumi/runtime/resource.py#L1089-L1113

`second` depends on `first`. `first` is a component resource so we don't add a dependency to `first` itself, instead, each of `first`'s childred is added as a dependency. `first` has one child: `second`. `second` is a component resource, so it isn't added as a dependency, it's children are. `second` has one child: `myresource`, a custom resource. Since `myresource` is a custom resource, it is added to the set of dependencies. We try to await its URN, but we'll never get it because `RegisterResource` hasn't yet been called for it. And we hang waiting.

One fix is to add a check: If the resource is a component resource and it's the same resource we were originally looking for dependencies from, then exit the expansion instead of looking at your own children.

**Comment by justinvp** [comment]
Notably: We used to not add transitive component dependencies in the Python SDK. This was added in #7732 and released in v3.10.1, which likely explains why this hang didn't exist previously when used with v2.x of the Python SDK.
